File: behavioral/iterator/iterator.java
import java.util.ArrayList;

public class IteratorPattern {
    public static class Cat {
        private String name;
        private String breed;

        public Cat(String name, String breed) {
            this.name = name;
            this.breed = breed;
        }

        public String getDescripion() {
            return String.format("%s cat %s.", breed, name);
        }
    }

    public interface Iterator {
        public boolean hasNext();
        public Object next();
    }

    public static class CatGangIterator implements Iterator {
        private ArrayList<Cat> cats;
        private int position = 0;

        public CatGangIterator(ArrayList<Cat> cats) {
            this.cats = cats;
        }

        public boolean hasNext() {
            if (position >= cats.size()) {
                return false;
            } else {
                return true;
            }
        }

        public Object next() {
            Cat cat = cats.get(position);
            position += 1;
            return cat;
        }
    }

    public static class CatGang {
        private ArrayList<Cat> cats = new ArrayList<Cat>();

        public void addCat(Cat cat) {
            cats.add(cat);
        }

        public Iterator createIterator() {
            return new CatGangIterator(cats);
        }
    }

    public static void main(String[] args) {
        Cat simba = new Cat("Simba", "Siamese");
        Cat oscar = new Cat("Oscar", "Maine Coon");
        Cat oliver = new Cat("Oliver", "British Shorthair");

        CatGang catGang = new CatGang();
        catGang.addCat(simba);
        catGang.addCat(oscar);
        catGang.addCat(oliver);

        Iterator catGangIterator = catGang.createIterator();

        while (catGangIterator.hasNext()) {
            Cat cat = (Cat)catGangIterator.next();
            System.out.println(cat.getDescripion());
        }
    }
}

File: behavioral/iterator/iterator.ts
class Cat {
    private _name: string;
    private _breed: string;

    constructor(name: string, breed: string) {
        this._name = name;
        this._breed = breed;
    }

    public getDescripion(): string {
        return `${this._breed} cat ${this._name}.`;
    }
}

interface IIterator {
    hasNext(): boolean;
    next(): Object;
}

class CatGangIterator implements IIterator {
    private _cats: Cat[];
    private _position: number = 0;

    constructor(cats: Cat[]) {
        this._cats = cats;
    }

    public hasNext(): boolean {
        if (this._position >= this._cats.length) {
            return false;
        } else {
            return true;
        }
    }

    public next(): Object {
        return this._cats[this._position++];
    }
}

class CatGang {
    private _cats: Cat[] = [];

    public addCat(cat: Cat): void {
        this._cats.push(cat);
    }

    public createIterator(): IIterator {
        return new CatGangIterator(this._cats);
    }
}

let simba: Cat = new Cat('Simba', 'Siamese');
let oscar: Cat = new Cat('Oscar', 'Maine Coon');
let oliver: Cat = new Cat('Oliver', 'British Shorthair');

let catGang: CatGang = new CatGang();
catGang.addCat(simba);
catGang.addCat(oscar);
catGang.addCat(oliver);

let catGangIterator: IIterator = catGang.createIterator();

while (catGangIterator.hasNext()) {
    let cat: Cat = <Cat>catGangIterator.next();
    console.log(cat.getDescripion());
}
